tools/src/test/java/org/apache/kafka/tools/other/ReplicationQuotasTestRig.java

```diff
@@ -0,0 +1,482 @@
+                    LOGGER.info("waiting... Leader rate on 101 is " + leaderRate);
+                record(leaderRates, broker.config().brokerId(), leaderRate);
+                if (leaderRate > 0)
+                    LOGGER.trace("Leader Rate on " + broker.config().brokerId() + " is " + leaderRate);
+
+                double followerRate = measuredRate(broker, QuotaType.FollowerReplication$.MODULE$);
+                record(followerRates, broker.config().brokerId(), followerRate);
+                if (followerRate > 0)
+                    LOGGER.trace("Follower Rate on " + broker.config().brokerId() + " is " + followerRate);
+            }
+        }
+
+        private double measuredRate(KafkaServer broker, QuotaType repType) {
+            MetricName metricName = broker.metrics().metricName("byte-rate", repType.toString());
+            return broker.metrics().metrics().containsKey(metricName)
+                ? (double) broker.metrics().metrics().get(metricName).metricValue()
+                : -1d;
+        }
+
+        String json(String... topic) {
+            String topicStr = Arrays.stream(topic).map(t -> "{\"topic\": \"" + t + "\"}").collect(Collectors.joining(","));
+            return "{\"topics\": [" + topicStr + "],\"version\":1}";
+        }
+
+        KafkaProducer<byte[], byte[]> createProducer(String brokerList) {
+            return TestUtils.createProducer(
+                brokerList,
+                0,
+                60 * 1000L,
+                1024L * 1024L,
+                Integer.MAX_VALUE,
+                30 * 1000,
+                0,
+                16384,
+                "none",
+                20 * 1000,
+                SecurityProtocol.PLAINTEXT,
+                Option.empty(),
+                Option.empty(),
+                new ByteArraySerializer(),
+                new ByteArraySerializer(),
+                false
+            );
+        }
+
+        Properties createBrokerConfig(int brokerId, String zkConnect) {
+            return TestUtils.createBrokerConfig(
+                brokerId,
+                zkConnect,
+                false, // shorten test time
+                true,
+                TestUtils.RandomPort(),
+                Option.empty(),
+                Option.empty(),
+                Option.empty(),
+                true,
+                false,
+                TestUtils.RandomPort(),
+                false,
+                TestUtils.RandomPort(),
+                false,
+                TestUtils.RandomPort(),
+                Option.empty(),
+                3,
+                false,
+                1,
+                (short) 1,
+                false);
+        }
+    }
+
+    static class Journal {
+        File log = new File(DIR, "Log.html");
+
+        public Journal() {
+            header();
+        }
+
+        void appendToJournal(ExperimentDef config) {
+            DecimalFormat format = new DecimalFormat("###,###.###");
```

Comment:
What about to make `private static final DecimalFormat DEC_FORMAT` const?

Reply:
No, `DecimalFormat` not synchronized.